File: backend/src/main/java/com/unina/ratatouille23/ordinazione/entity/Ordinazione.java
package com.unina.ratatouille23.ordinazione.entity;

import java.util.Date;
import java.util.List;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.unina.ratatouille23.utente.entity.Utente;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Temporal;
import jakarta.persistence.TemporalType;

@Entity
public class Ordinazione {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id_ordinazione")
    private Integer id;

    @Column(name = "id_ristorante", nullable = false)
    private int idRistorante;

    @ManyToOne
    @JoinColumn(name = "fk_id_gestore")
    @JsonIgnoreProperties(value = {"applications", "hibernateLazyInitializer"})
    @JsonInclude(Include.NON_NULL)
    private Utente gestoreOrdinazione;

    @ManyToOne
    @JoinColumn(name = "fk_id_addettosala")
    @JsonIgnoreProperties(value = {"applications", "hibernateLazyInitializer"})
    private Utente addettoAllaSala;

    @Column(name = "n_tavolo", nullable = false)
    private int numeroTavolo;

    private String note;

    @Column(nullable = false)
    private String stato;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date data; 

    @Column(nullable = false)
    private double costoTotale;

    @OneToMany(mappedBy = "ordinazione", targetEntity = ElementoOrdinato.class, cascade = CascadeType.MERGE, fetch = FetchType.LAZY, orphanRemoval = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonManagedReference
    // @JoinColumn(name = "fk_id_ordinazione", referencedColumnName = "id_ordinazione")
    //Owner della relazione, viene istanziata l'ordinazione e poi vengono istanziati gli elementi ordinati
    //con l'id dell'ordinazione come foreign key
    private List<ElementoOrdinato> elementiOrdinati;

    

    public Ordinazione(int idRistorante, Integer id, Utente gestoreOrdinazione, int numeroTavolo, String note, String stato, Date data, double costoTotale, List<ElementoOrdinato> elementiOrdinati) {
        this.id = id;
        this.idRistorante = idRistorante;
        this.gestoreOrdinazione = gestoreOrdinazione;
        this.numeroTavolo = numeroTavolo;
        this.note = note;
        this.stato = stato;
        this.data = data;
        this.costoTotale = costoTotale;
        this.elementiOrdinati = elementiOrdinati;
    }


    public Ordinazione() {
    }


    public int getIdRistorante() {
        return this.idRistorante;
    }

    public void setIdRistorante(int idRistorante) {
        this.idRistorante = idRistorante;
    }

    
    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public int getNumeroTavolo() {
        return numeroTavolo;
    }

    public void setNumeroTavolo(int numeroTavolo) {
        this.numeroTavolo = numeroTavolo;
    }

    public String getNote() {
        return note;
    }

    public void setNote(String note) {
        this.note = note;
    }

    public String getStato() {
        return stato;
    }

    public void setStato(String stato) {
        this.stato = stato;
    }

    public Date getData() {
        return data;
    }

    public void setData(Date data) {
        this.data = data;
    }

    public double getCostoTotale() {
        return costoTotale;
    }

    public void setCostoTotale(double costoTotale) {
        this.costoTotale = costoTotale;
    }

    public List<ElementoOrdinato> getElementiOrdinati() {
        return elementiOrdinati;
    }

    public void setElementiOrdinati(List<ElementoOrdinato> elementiOrdinati) {
        this.elementiOrdinati = elementiOrdinati;
    }

    public Utente getGestoreOrdinazione() {
        return gestoreOrdinazione;
    }

    public void setGestoreOrdinazione(Utente gestoreOrdinazione) {
        this.gestoreOrdinazione = gestoreOrdinazione;
    }

    public Utente getAddettoAllaSala() {
        return addettoAllaSala;
    }

    public void setAddettoAllaSala(Utente addettoAllaSala) {
       this.addettoAllaSala =  addettoAllaSala;
    }
    

}

File: client/lib/models/Ordinazione.dart

import 'package:intl/intl.dart';
import 'package:ratatouille23/models/Elemento_ordinato.dart';
import 'package:ratatouille23/models/Utente.dart';

import 'menu/Elemento.dart';

class Ordinazione{
  late int _id;
  late int _tavolo;
  late String _note;
  late String _stato;
  late DateTime _data;
  late List<Elemento_ordinato> _elementi;
  late double _costo_totale;
  Utente? _gestore_ordinazione;
  Utente? _addetto_alla_sala;


  double _calcola_totale(List<Elemento_ordinato> elementi){
    double totale=0;
    for(int i=0;i<elementi.length;i++){
      totale=totale+elementi[i].get_elemento().costo;
    }
    return totale;
  }

  Ordinazione.senzaId(int tavolo, String note, List<Elemento_ordinato> elementi, Utente addetto_alla_sala){
    _tavolo=tavolo;
    _note=note;
    _stato='In attesa';
    _data= DateTime.now();
    _elementi=elementi;
    _costo_totale=_calcola_totale(elementi);
    _addetto_alla_sala=addetto_alla_sala;
  }

    Ordinazione.fromJson(Map<String, dynamic> json){
      _id = json['id'];
      _tavolo=json['numeroTavolo'];
      _note=json['note'];
      _stato=json['stato'];
      _data=DateTime.parse(json['data']);
      _costo_totale=json['costoTotale'];
      _gestore_ordinazione=Utente.fromJson(json['gestoreOrdinazione']);
      _addetto_alla_sala=Utente.fromJson(json['addettoAllaSala']);
      _elementi = [];
      json['elementiOrdinati'].forEach((v) {
        _elementi.add(Elemento_ordinato.fromJson(v));
      });
  }

  Ordinazione.fromJsonSenzaGestore(Map<String, dynamic> json){
    _id = json['id'];
    _tavolo=json['numeroTavolo'];
    _note=json['note'];
    _stato=json['stato'];
    _data=DateTime.parse(json['data']);
    _costo_totale=json['costoTotale'];
    _addetto_alla_sala=Utente.fromJson(json['addettoAllaSala']);
    _elementi = [];
    json['elementiOrdinati'].forEach((v) {
      _elementi.add(Elemento_ordinato.fromJson(v));
    });
  }

  Ordinazione.fromJsonSenzaAddettoAllaSala(Map<String, dynamic> json){
    _id = json['id'];
    _tavolo=json['numeroTavolo'];
    _note=json['note'];
    _stato=json['stato'];
    _data=DateTime.parse(json['data']);
    _costo_totale=json['costoTotale'];
    _gestore_ordinazione=Utente.fromJson(json['gestoreOrdinazione']);
    _elementi = [];
    json['elementiOrdinati'].forEach((v) {
      _elementi.add(Elemento_ordinato.fromJson(v));
    });
  }

  Map<String, dynamic> toJson(){
      List<Map> jsonElem = _elementi.map((e) => e.toJson()).toList();
      return{
        'idRistorante':_addetto_alla_sala?.get_id_ristorante(),
        'id':_id,
        'numeroTavolo':_tavolo,
        'note':_note,
        'stato':_stato,
        'data':fromDateToString(),
        'elementiOrdinati':jsonElem,
        'costoTotale':_costo_totale,
        'gestoreOrdinazione':_gestore_ordinazione,
        'addettoAllaSala':_addetto_alla_sala
      };
    }

  Map<String, dynamic> toJsonSenzaId(){
    return{
      'tavolo':_tavolo,
      'note':_note,
      'stato':_stato,
      'data':fromDateToString(),
      'elementi':_elementi,
      'costo_totale':_costo_totale,
      'gestore_ordinazione':_gestore_ordinazione
    };
  }

  Map<String, dynamic> toJsonSenzaAddettoAllaSala(){
    List<Map> jsonElem = _elementi.map((e) => e.toJson()).toList();
    return{
      'id':_id,
      'numeroTavolo':_tavolo,
      'note':_note,
      'stato':_stato,
      'data':fromDateToString(),
      'elementiOrdinati':jsonElem,
      'costoTotale':_costo_totale,
      'gestoreOrdinazione':_gestore_ordinazione
    };
  }

  Map<String, dynamic> toJsonSenzaGestore(){
    List<Map> jsonElem = _elementi.map((e) => e.toJson()).toList();
    return{
      'idRistorante':_addetto_alla_sala?.get_id_ristorante(),
      'numeroTavolo':_tavolo,
      'note':_note,
      'stato':_stato,
      'data':fromDateToString(),
      'elementiOrdinati':jsonElem,
      'costoTotale':_costo_totale,
      'addettoAllaSala':_addetto_alla_sala,
    };
  }

  String fromDateToString() {
    var formatter = new DateFormat('yyyy-MM-dd');
    String formattedDate = formatter.format(_data);
    return formattedDate;
  }

    void set_id(int id){_id=id;}
    void set_tavolo(int tavolo){_tavolo=tavolo;}
    void set_note(String note){_note=note;}
    void set_stato(String stato){_stato=stato;}
    void set_lista_elementi(List<Elemento_ordinato> elementi){_elementi=elementi;}
    void set_gestore_ordinazione(Utente gestore_ordinazione){_gestore_ordinazione=gestore_ordinazione;}

    int get_id(){return _id;}
    int get_tavolo(){return _tavolo;}
    String get_note(){return _note;}
    String get_stato(){return _stato;}
    DateTime get_data(){return _data;}
    List<Elemento_ordinato> get_lista_elementi(){return _elementi;}
    double get_costo_totale(){return _costo_totale;}
    Utente? get_gestore_ordinazione(){return _gestore_ordinazione;}
    Elemento get_elemento(int i){return _elementi[i].get_elemento();}
    int get_quantita_elemento(int i){return _elementi[i].get_quantita();}

}